lib/Dialect/LLHD/Simulator/State.cpp

```diff
@@ -101,49 +101,135 @@ bool Slot::operator<(const Slot &rhs) const { return time < rhs.time; }
+    buffers[changesSize] = std::make_pair(
+        bitOffset,
+        APInt(width, makeArrayRef(reinterpret_cast<uint64_t *>(bytes), size)));
+  }
+
+  // Map the signal index to the change buffer so we can retrieve
+  // it after sorting.
+  changes.push_back(std::make_pair(index, changesSize));
+  ++changesSize;
 }
 
-void Slot::insertChange(std::string inst) { scheduled.push_back(inst); }
+void Slot::insertChange(unsigned inst) { scheduled.push_back(inst); }
 
 //===----------------------------------------------------------------------===//
 // UpdateQueue
 //===----------------------------------------------------------------------===//
 void UpdateQueue::insertOrUpdate(Time time, int index, int bitOffset,
-                                 APInt &bytes) {
-  for (size_t i = 0, e = c.size(); i < e; ++i) {
-    if (time == c[i].time) {
-      c[i].insertChange(index, bitOffset, bytes);
-      return;
-    }
-  }
-  Slot newSlot(time);
-  newSlot.insertChange(index, bitOffset, bytes);
-  push(newSlot);
+                                 uint8_t *bytes, unsigned width) {
+  auto &slot = getOrCreateSlot(time);
+  slot.insertChange(index, bitOffset, bytes, width);
 }
 
-void UpdateQueue::insertOrUpdate(Time time, std::string inst) {
-  for (size_t i = 0, e = c.size(); i < e; ++i) {
-    if (time == c[i].time) {
-      c[i].insertChange(inst);
-      return;
+void UpdateQueue::insertOrUpdate(Time time, unsigned inst) {
+  auto &slot = getOrCreateSlot(time);
+  slot.insertChange(inst);
+}
+
+Slot &UpdateQueue::getOrCreateSlot(Time time) {
+  auto &top = begin()[topSlot];
+
+  // Directly add to top slot.
+  if (!top.unused && time == top.time) {
+    return top;
+  }
+
+  // We need to search through the queue for an existing slot only if we're
+  // spawning an event later than the top slot. Adding to an existing slot
+  // scheduled earlier than the top slot should never happens, as then it should
+  // be the top.
+  if (events > 0 && top.time < time) {
+    for (size_t i = 0, e = size(); i < e; ++i) {
+      if (time == begin()[i].time) {
+        return begin()[i];
+      }
     }
   }
-  Slot newSlot(time);
-  newSlot.insertChange(inst);
-  push(newSlot);
+
+  // Spawn new event using an existing slot.
+  if (unused.size() > 0) {
+    auto firstUnused = unused.back();
+    unused.pop_back();
+    auto &newSlot = begin()[firstUnused];
+    newSlot.unused = false;
+    newSlot.time = time;
+
+    // Update the top of the queue either if it is currently unused or the new
+    // timestamp is earlier than it.
+    if (top.unused || time < top.time)
+      topSlot = firstUnused;
+
+    ++events;
+    return newSlot;
+  }
+
+  // We do not have pre-allocated slots available, generate a new one.
+  push_back(Slot(time));
+
+  // Update the top of the queue either if it is currently unused or the new
+  // timestamp is earlier than it.
+  if (top.unused || time < top.time)
+    topSlot = size() - 1;
+
+  ++events;
+  return back();
+}
```

Comment:
Very nice. This looks like it will make the common path fast, and the uncommon path slower. Very nice.